src/process/fork.rs

```diff
@@ -1,15 +1,84 @@
 use anyhow::Result;
+use libc::c_int;
+use libc::c_void;
+use nix::errno::Errno;
 use nix::sched;
+use nix::sys;
 use nix::unistd::Pid;
+use std::mem;
 
-pub fn clone(cb: sched::CloneCb, clone_flags: sched::CloneFlags) -> Result<Pid> {
-    // unlike fork, clone requires the caller to allocate the stack. here, we use the default
-    // 4KB for stack size, consistant with the runc implementation.
-    const STACK_SIZE: usize = 4096;
-    let stack: &mut [u8; STACK_SIZE] = &mut [0; STACK_SIZE];
-    // pass in the SIGCHID flag to mimic the effect of forking a process
-    let signal = nix::sys::signal::Signal::SIGCHLD;
-    let pid = sched::clone(cb, stack, clone_flags, Some(signal as i32))?;
+/// clone uses syscall clone(2) to create a new process for the container init
+/// process. Using clone syscall gives us better control over how to can create
+/// the new container process, where we can enter into namespaces directly instead
+/// of using unshare and fork. This call will only create one new process, instead
+/// of two using fork.
+pub fn clone(mut cb: sched::CloneCb, clone_flags: sched::CloneFlags) -> Result<Pid> {
+    extern "C" fn callback(data: *mut sched::CloneCb) -> c_int {
+        let cb: &mut sched::CloneCb = unsafe { &mut *data };
+        (*cb)() as c_int
+    }
+
+    // Use sysconf to find the page size. If there is an error, we assume
+    // the default 4K page size.
+    let page_size: usize = unsafe {
+        match libc::sysconf(libc::_SC_PAGE_SIZE) {
+            -1 => 4 * 1024, // default to 4K page size
+            x => x as usize,
+        }
+    };
+
+    // Find out the default stack max size through getrlimit.
```

Comment:
Could you comment that we are waiting for this PR?
https://github.com/nix-rust/nix/pull/1302
I'd like to avoid direct use of libc if possible.

Reply:
There is another crate for getrlimit.
Its design is different from nix.
https://docs.rs/rlimit/0.6.2/rlimit/

Reply:
@utam0k Yes, that PR would provide the nix wrapper for `getrlimit`, which is not there yet. Another way is we use the rlimit crate mentioned here just for this. Your call.

Reply:
> There is another crate for getrlimit.
> Its design is different from nix.
> https://docs.rs/rlimit/0.6.2/rlimit/

Thank you for the wonderful crate introduction. This time, nix is starting to support rlimit, so I'll wait for that.

Reply:
> @utam0k Yes, that PR would provide the nix wrapper for `getrlimit`, which is not there yet. Another way is we use the rlimit crate mentioned here just for this. Your call.

This implementation is fine for this time, but please comment in your code that you are waiting for this PR.

Reply:
So let's leave the `libc::getrlimit` in this PR and create another issue to track when nix supports `getrlimit`?

Reply:
That's a good idea. Let's do that.

Reply:
#168 filed.